tooling/bundler/src/bundle/windows/templates/installer.nsi

```diff
@@ -702,6 +708,7 @@ Section Uninstall
     !insertmacro UnpinShortcut "$SMPROGRAMS\$AppStartMenuFolder\${PRODUCTNAME}.lnk"
     Delete "$SMPROGRAMS\$AppStartMenuFolder\${PRODUCTNAME}.lnk"
     RMDir "$SMPROGRAMS\$AppStartMenuFolder"
```

Comment:
I don't think we should call these anymore, unless the installer have created it, maybe we should check if `$AppStartMenuFolder` is not empty or something or if it equals `${STARTMENUFOLDER}`.

Reply:
There're no good way of knowing that to be honest, `$AppStartMenuFolder` returns default start menu folder if no registry key is found and also when that registry key is found and is the default value

Reply:
There're no good way of knowing that to be honest, `$AppStartMenuFolder` returns default start menu folder if no registry key is found and also when that registry key is found and is the default value

Reply:
yeah, I was afraid of that, maybe we should check if it is at least equal to `${STARTMENUFOLDER}`, that at least should be an indication to remove this.

Reply:
I don't see any downside of just try deleting this except the potential extra attempt to be honest

Reply:
yeah I guess, but avoiding deleting files that wasn't created by us still better than accidentally deleting a shortcut and a folder that may or may not be created by us

Reply:
I don't see a good way of detecting if it's created by us or not to be honest

Reply:
just checking if `$AppStartMenuFolder` is equal to `${STARTMENUFOLDER}` should be enough IMO. if there is residual files after that, we can't be blamed as that is our best effort to cleanup

Reply:
If it was `Tauri API/Tauri API.lnk`, and we changed `${STARTMENUFOLDER}` to `""`, `$AppStartMenuFolder` will not be equal to `${STARTMENUFOLDER}` and it doesn't get clean up

Reply:
When running the uninstaller, it will have the same values as the installer that created the uninstaller not the one trying to update, so the value of AppStartMenuFolder will be the same

Reply:
But it's no longer true after update if we changed `${STARTMENUFOLDER}` to `""`

Another way is to migrate old shortcuts on update, but we need to preserve the old start menu shortcuts because if we migrate them, pinned old shortcuts will be gone

Reply:
Well, I just realized that we didn't specify these

Reply:
we didn't store anything before, there're no way of knowing if we had these or not

after this change, we'll have most of tauri apps migrating to this and if we choose not to clean it up, then almost everyone will be unhappy about installing, if we delete it, there will be a few occasions where some other app with exactly the same name and folder name will get affected

Reply:
> But it's no longer true after update if we changed `${STARTMENUFOLDER}` to `""`

correct, but at that point, I don't think we should care about it

> Another way is to migrate old shortcuts on update, but we need to preserve the old start menu shortcuts because if we migrate them, pinned old shortcuts will be gone

That's not needed, it probably will be more headache

Reply:
> Well, I just realized that we didn't specify these
> 
> ![image](https://private-user-images.githubusercontent.com/68118705/339641665-136f91c9-e0ea-4696-80b1-05e81fd0b63f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNDQ1NTQsIm5iZiI6MTcxODM0NDI1NCwicGF0aCI6Ii82ODExODcwNS8zMzk2NDE2NjUtMTM2ZjkxYzktZTBlYS00Njk2LTgwYjEtMDVlODFmZDBiNjNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE0VDA1NTA1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhOTUxZDE0YTAyOGQzODQ3MDI5MzJiOWVhMDFjYThlOGE0NWEwNWFlYmZlOTRkZGU5ZjQ3ODg3ZmE1ZDQ3MzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hKfCvRklpBlcCxcDh3wNaPn-Hpc-aqap1yjcr6MrIjo)
> 
> Currently it can't even clean up start menu shortcuts if the user chose another folder

It doesn't use defaults? bummer, we should specify those then

Reply:
> We didn't store anything before, there're no way of knowing if we had these or not
> 
> After this change, we'll have most of tauri apps migrating to this and if we choose not to clean it up, then almost everyone will be unhappy about installing, if we delete it, there will be some really rare occasions where some other app with exactly the same name and folder name will get affected (they'll have some other problems like desktop shortcut anyway)

yes but on one hand, we are deleting, which is destructive and on the other we are being careful not to delete arbitrary files that we are not at least 50% percent sure we should delete. 

In ideal scenarios, the shortcut location will be the same between updates and nothing changes for the user.

In worst case, an extra shortcut will be leftover, not the end of the world and if the user stumbles upon a broken shortcut, Windows will prompt the user to delete it.

Reply:
I would argue that our app simply occupies both `ProductName.lnk` and `ProductName/ProductName.lnk`, if there is a conflict, then be it

Reply:
> I would argue that our app simply occupies both `ProductName.lnk` and `ProductName/ProductName.lnk`, if there is a conflict, then be it
> 
> > that we are not at least 50% percent sure we should delete
> 
> After this, I would expect it to be 99%+ on existing apps

yeah but on future apps it won't be, it will bet at least 50%

> > In worst case, an extra shortcut will be leftover, not the end of the world and if the user stumbles upon a broken shortcut, Windows will prompt the user to delete it.
> 
> Yes, but I would have a very bad impression about that app

It would leave a bad impression if your app deletes another shortcut though

> > It doesn't use defaults? bummer, we should specify those then
> 
> This thing is quite tricky to be honest, it depends on multiple things in different sections, I will leave to another day to fix, not this PR

No problem, thanks for helping

Reply:
Well, if you persist, I wouldn't fight on this, but is there anyway to mitigate this for old apps? I can just add a check here and only remove plain shortcuts if no `${STARTMENUFOLDER}` and only remove the one in the folder other wise

Reply:
> is there anyway to mitigate this for old apps? I can just add a check here and only remove plain shortcuts if no `${STARTMENUFOLDER}` and only remove the one in the folder other wise

I honestly hate this; we seem to be cycling to this same issue every time we make a breaking change like this, "can we migrate old app shortcut?" yes, sure but it comes at costs tbh. Ideally, we should be smart about deleting or updating existing shortcuts, I wonder if we can associate the app identifier in the shortcut properties, so that we can query them later to decide whether to update or delete a shortcut.

> > It would even be a worse impression if your app deletes another shortcut though
> 
> To be honest, if you really think about it, we try to re-target old shortcuts to our app, it's just as destructive as this (and happens a lot more frequently on very update), I don't see how we can deal with this type of conflicts anyway

that is indeed destructive as well, might as well delete them all then

Reply:
The problem about making changes on updaters is that we can't really make breaking changes, they're directly impacting the user, not like other breaking changes are just affecting the devs

> I wonder if we can associate the app identifier in the shortcut properties, so that we can query them later to decide whether to update or delete a shortcut

We probably can detect if the shortcut's target is on our binary or not, but I don't know how practical is it to be done in nsis

> that is indeed destructive as well, might as well delete them all then

Do you mean we just delete the shortcuts on uninstall or delete the migration code?

Reply:
I just pushed a commit to detect if the shortcut is actually pointing to our binary, this should prevent all those things